python/src/opendp/extras/sklearn/linear_model/_make_private_theil_sen/__init__.py

```diff
@@ -0,0 +1,119 @@
+import opendp.prelude as dp
+from opendp._lib import import_optional_dependency
+
+
+def pairwise_predict(data, x_cuts):
+    """
+    The function matches data points together into pairs,
+    fits a line that passes through each pair of points,
+    and then computes the y values of each line at `x_cuts`.
+    That is, it predicts the y values at `x_cuts`, pairwise.
+    """
+    np = import_optional_dependency("numpy")
+    # Get an even number of rows.
+    data = np.array(data, copy=True)[: len(data) // 2 * 2]
+
+    # Shuffle the data to get random matchings.
+    np.random.shuffle(data)
+
+    # Split data into pairs, where pair i is (p1[i], p2[i]).
+    p1, p2 = np.array_split(data, 2)
+
+    # Compute differences.
+    dx, dy = (p2 - p1).T
+
+    # Compute the midpoints of the pairs.
+    x_bar, y_bar = (p1 + p2).T / 2
+
+    # Compute y values on the pairwise slopes at x_cuts.
+    points = dy / dx * (x_cuts[None].T - x_bar) + y_bar
+
+    # Pairs where the x difference is zero are degenerate.
+    return points.T[dx != 0]
+
+
+def make_pairwise_predict(x_cuts, runs: int = 1):
+    """
+    The parameter `runs` controls how many times randomized pairwise predictions are computed.
+    The default is 1. Increasing `runs` can improve the robustness and accuracy of the results;
+    however, it can also increase computational cost and amount of noise needed later in the algorithm.
+    """
+    np = import_optional_dependency("numpy")
+    return dp.t.make_user_transformation(
+        # Outputs are Nx2 non-nan float numpy arrays.
+        input_domain=dp.numpy.array2_domain(num_columns=2, T=float),
```

Comment:
Can T be an argument that defaults to `float`?

Reply:
The comment is wrong, it doesn't need to be non-nan.

Reply:
done. (Just changed the type param here, not in notebook.)